pilot/pkg/config/coredatamodel/controller.go

```diff
@@ -70,61 +71,92 @@ func (c *Controller) ConfigDescriptor() model.ConfigDescriptor {
 }
 
 // List returns all the config that is stored by type and namespace
-func (c *Controller) List(typ, namespace string) ([]model.Config, error) {
+// if namespace is empty string it returns config for all the namespaces
+func (c *Controller) List(typ, namespace string) (out []model.Config, err error) {
 	_, ok := c.ConfigDescriptor().GetByType(typ)
 	if !ok {
 		return nil, fmt.Errorf("list unknown type %s", typ)
 	}
 	data, exists := c.configStore[typ]
-	if !exists {
-		log.Infof("list config not found for the type %s\n", typ)
-		return nil, nil
+	if exists && len(data) == 0 {
+		return out, nil
 	}
-	out := make([]model.Config, 0, len(c.configStore[typ]))
+
 	if namespace == "" {
-		for _, ns := range data {
-			ns.Range(func(key, value interface{}) bool {
-				out = append(out, value.(model.Config))
-				return true
-			})
+		c.configStoreMu.Lock()
+		byType, ok := c.configStore[typ]
+		c.configStoreMu.Unlock()
+		if ok {
+			for _, byNamespace := range byType {
+				for _, config := range byNamespace {
+					out = append(out, config)
+				}
+			}
+			return out, nil
 		}
-	} else {
-		ns, exists := data[namespace]
-		if !exists {
-			return nil, nil
+	}
+
+	if byNamespace, ok := data[namespace]; ok {
+		for _, config := range byNamespace {
+			out = append(out, config)
 		}
-		ns.Range(func(key, value interface{}) bool {
-			out = append(out, value.(model.Config))
-			return true
-		})
 	}
 	return out, nil
 }
 
 // Apply receives changes from MCP server and creates the
 // corresponding config
 func (c *Controller) Apply(change *mcpclient.Change) error {
+	typ := extractType(change.MessageName)
```

Comment:
`change.MessageName` is the protobuf message name without the `types.googleapis.com` prefix. You don't need to strip anything (yet). 

Note that this will change once https://github.com/istio/istio/pull/8207 is merged.

Reply:
Should we leave it in place since it handles either cases?

Reply:
You could leave it as a no-op for now since we know we'll be changed this soon. In either case, I would rename to `extractMessageName` since we'll be going from `TypeUrl` to `MessageName`.

Reply:
Updated, removed it to avoid confusion for now, it's easy to throw back in there once #8207 is merge.